analyzers/src/SonarAnalyzer.CSharp/Rules/InfiniteRecursion_RoslynCfg.cs

```diff
@@ -0,0 +1,224 @@
+                            cfg = cfg.GetAnonymousFunctionControlFlowGraph(operationWrapper);
+                        }
+                    }
+
+                    cfg = cfg.GetLocalFunctionControlFlowGraph(symbol as IMethodSymbol);
+                }
+
+                var walker = new CfgWalkerForMethod(new RecursionAnalysisContext<ControlFlowGraph>(cfg, symbol, identifier.GetLocation(), c));
+                if (!walker.CheckPaths() && !cfg.ExitBlock.IsReachable)
+                {
+                    c.ReportDiagnosticWhenActive(Diagnostic.Create(Rule, identifier.GetLocation(), "method's recursion"));
+                }
+            }
+
+            private class CfgWalkerForMethod : CfgRecursionSearcher
+            {
+                private readonly BooleanBlockMatcher existReferenceToDeclaringSymbolMatcher;
+
+                public CfgWalkerForMethod(RecursionAnalysisContext<ControlFlowGraph> context)
+                    : base(
+                        context.ControlFlowGraph,
+                        context.AnalyzedSymbol,
+                        context.SemanticModel,
+                        () => context.AnalysisContext.ReportDiagnosticWhenActive(Diagnostic.Create(Rule, context.IssueLocation, "method's recursion")))
+                {
+                    existReferenceToDeclaringSymbolMatcher = new BooleanBlockMatcher(operation =>
+                    {
+                        if (!(operation.Instance.Syntax is InvocationExpressionSyntax invocation))
+                        {
+                            return false;
+                        }
+
+                        return IsInstructionOnThisAndMatchesDeclaringSymbol(invocation.Expression, declaringSymbol, semanticModel);
+                    });
+                }
+
+                protected override bool BlockHasReferenceToDeclaringSymbol(BasicBlock block) =>
+                    existReferenceToDeclaringSymbolMatcher.Matches(block);
+            }
+
+            private class CfgWalkerForProperty : CfgRecursionSearcher
+            {
+                private readonly bool isSet;
+                private readonly BooleanBlockMatcher existReferenceToDeclaringSymbolMatcher;
+
+                public CfgWalkerForProperty(RecursionAnalysisContext<ControlFlowGraph> context, string reportOn, bool isSetAccessor)
+                    : base(
+                        context.ControlFlowGraph,
+                        context.AnalyzedSymbol,
+                        context.SemanticModel,
+                        () => context.AnalysisContext.ReportDiagnosticWhenActive(Diagnostic.Create(Rule, context.IssueLocation, reportOn)))
+                {
+                    isSet = isSetAccessor;
+                    existReferenceToDeclaringSymbolMatcher = new BooleanBlockMatcher(operation =>
+                        operation.Instance.Syntax is var syntax
+                        && TypesForReference.Contains(syntax.GetType())
```

Comment:
This looks heavy. Can we use `.IsAnyKind()` instead?

Reply:
Nope, since MemberAccessExpressionSyntax has no matching syntax kind

Reply:
It has. `SyntaxKind.SimpleMemberAccessExpression` and `SyntaxKind.PointerMemberAccessExpression` as seen here
https://github.com/dotnet/roslyn/blob/0700dcde657135078bf8c869129a64a266e110cf/src/Compilers/CSharp/Portable/Syntax/Syntax.xml#L556-L559

Reply:
But I'm resolving this, as it went away anyway.